src/v/archival/probe.cc

```diff
@@ -105,23 +129,57 @@ void ntp_level_probe::setup_public_metrics(const model::ntp& ntp) {
            "This may grow due to retention or non compacted segments being "
            "replaced with their compacted equivalent."),
          labels)
-         .aggregate(aggregate_labels),
-       sm::make_gauge(
-         "segments",
-         [this] { return _segments_in_manifest; },
-         sm::description(
-           "Total number of accounted segments in the cloud for the topic"),
-         labels)
-         .aggregate(aggregate_labels),
-       sm::make_gauge(
-         "segments_pending_deletion",
-         [this] { return _segments_to_delete; },
-         sm::description("Total number of segments pending deletion from the "
-                         "cloud for the topic"),
-         labels)
          .aggregate(aggregate_labels)});
 }
 
+void ntp_level_probe::setup_dynamic_public_metrics(const model::ntp& ntp) {
+    namespace sm = ss::metrics;
+
+    auto ns_label = ssx::metrics::make_namespaced_label("namespace");
+    auto topic_label = ssx::metrics::make_namespaced_label("topic");
+    auto partition_label = ssx::metrics::make_namespaced_label("partition");
+    const std::vector<sm::label_instance> labels = {
+      ns_label(ntp.ns()),
+      topic_label(ntp.tp.topic()),
+      partition_label(ntp.tp.partition()),
+    };
+
+    auto aggregate_labels = std::vector<sm::label>{
+      sm::shard_label, partition_label};
+
+    _dynamic_public_metrics->add_group(
+      prometheus_sanitize::metrics_name("cloud_storage"),
+      {
+        sm::make_gauge(
+          "segments",
+          [this] { return _segments_in_manifest; },
+          sm::description(
+            "Total number of accounted segments in the cloud for the topic"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_gauge(
+          "segments_pending_deletion",
+          [this] { return _segments_to_delete; },
+          sm::description("Total number of segments pending deletion from the "
+                          "cloud for the topic"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_gauge(
+          "records_pending_upload",
+          [this] { return _parent.records_pending_upload(); },
+          sm::description("The number of records in the local log that have "
+                          "not yet been uploaded"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_gauge(
+          "estimated_backlog_size_bytes",
+          [this] { return _parent.estimate_backlog_size(); },
```

Comment:
estimate_backlog_size is a for loop over local segments, so this is probably too expensive to hit on every prometheus scrape